Add support for more widgets and improve activation

rwx788 opened the pull request:
For many widgets we have different behavior in qt and ncurses, so we
were adding activation methods to the libyui ncurses and qt libraries.
But as we already did the work to split libyui-rest-api in 3 packages,
we can override implementations for some widgets only.

Also, this PR makes json return format consistent, especially in case of error handling.

Some code duplication was removed by using templates as widgets are handled identically in many cases.

Dependent pull requests:
- [libyui](https://github.com/libyui/libyui/pull/166)
- [libyui-qt-rest-api](https://github.com/libyui/libyui-qt-rest-api/pull/4)
- [libyui-ncurses-rest-api](https://github.com/libyui/libyui-ncurses-rest-api/pull/3)

Commit:
Activate UI when item is selected in combobox

We had a bug in the implementation and selection of the item in combobox
didn't trigger UI update. That was visible in the yast2 timezone module
in qt. Tested for both ncurses and qt now.

Commit:
Trigger UI update when select in YSelectionBox

Commit:
Add support for YMultiSelectionBox for rest api

Commit:
Use json as content type and not encoding

This commit fixes parameter to set response header to json format and
not encoding, as we don't actually use it as of now. Also, this commit
fixes all the cases where not json was returned as an error message to
be consistent.

Commit:
Do early return in case of errors

In case we don't find any widgets or find multiple by given criteria, we
should not try to perform any action.

Commit:
Add support for YDateField

Commit:
Add support for Date and Time fields

Commit:
Allow multi criteria search for the reading the widgets

We have introduced that for POST requests, but not for the GET
requests.

Commit:
Do early return if no search criteria provided for action

In case there is no search criteria we can safely do early return
because it's unlikely that there will be single widget in the app and we
don't support actions on multiple controls simultaneously.

Commit:
Move YRichText and YMenuButton to select action

YRichText and YMenuButton both has "press" action specified in
YHttpWidgetsActionHandler.

The problem is they also require 'value' to be send, though YPushButton
only requires 'press' action.

With select it will be more consistent.

Commit:
Allow overrides for widget action handler

For many widgets we have different behavior in qt and ncurses, so we
were adding activation methods to the libyui ncurses and qt libraries.
But as we already did the work to split libyui-rest-api in 3 packages,
we can override implementations for some widgets.
This required removing logging from the YHttpWidgetsActionHandler.h
header file, which is fine for now and will be addressed later, as
logging is used in the template methods.

Commit:
Introduce activation methods to be overriden

Instead of override for the whole do_action method, we actually just
need to handle triggering of the events differently, so introducing the
structure which allows to override activation for particular widgets in
ncurses and qt libraries accordingly.

Commit:
Add activation for YInputField

We need activation for qt only, in ncurses UI is updated without
additional event. So we add empty implementation which will be used in
case of missing override.
This commit also contains small improvements to the header file,
optimizing imports and moving methods implementations to the cpp file.

Commit:
Move YMultiSelectionBox activation to nc and qt rest-api libraries

Commit:
Add support to enter value in editable comboboxes

Commit:
Return error if try to apply action on disabled widget

To provide more consistent errors, we should not return OK code when
client attempted to activate disabled element.
This is also required for test execution sync as we might wait for
widget to get enabled to apply an action, so we can retry multiple times
till we receive OK code. Otherwise we will have to check for the result
of action without knowing if it was triggered or not.

rwx788 commented:
Once #4 is merged, I will have to resolve conflicts and will update the documentation to describe current state of the things.

Commit:
Add support to select row in the table by id

We might need to select row in the table simply by id. Now it's possible
with the request using parameter `row`.

For example:
`http://localhost:9999/widgets?type=YTable&action=select&row=1`

lslezak commented on src/YHttpWidgetsActionHandler.h:
```diff
@@ -46,45 +51,108 @@ class YHttpWidgetsActionHandler : public YHttpHandler
             {
-                yuiError() << "Error while processing action on widget "
-                    << typeid(*widget).name() << " " << e.what() << std::endl;
+                body << "{ error: \"" << typeid(*widget).name() << " " << e.what() << "\" }" << std::endl;
```
Using a literal JSON is fine for static error messages:
```cpp
body << "{ \"error\" : \"Widget not found\" }" << std::endl;
```

But for variable massages which can potentially contain some user input it is not safe. E.g. if the exception text contains a double quote, new line or other special characters then it can break the JSON format. In these cases we would need to escape the message text. But it's easier to simply use the JSON library for that:

```cpp
Json::Value response;
response["error"] = typeid(*widget).name() + " " + e.what();
YJsonSerializer::save(response, body);
```

rwx788 replied:
Makes sense, I will fix that while resolving conflicts. Thanks!

Commit:
Use json serializer for building complex error messages

During the review Ladi has correctly pointed out that in case we build
complex error string, it's easy to break json format, so using json
serializer will make this less error prone.

Commit:
Update project and rest api documentation

rwx788 commented on README.md:
```diff
@@ -76,90 +97,61 @@ in clear text without any encryption (only converted to the Base64 encoding).
+For running those examples `yast2-core` package has to be preinstalled.
+
+After `yast-ycp-ui-bindings` project is checked out or `yast2-ycp-ui-bindings-devel`
+package is installed, copy all the files to the `clients` subdirectory anywhere
```
@lslezak @jreidinger Hey! I was not sure if there is better way to run examples from that repo. If not, I would propose to create `clients` directory in the examples, so this extra step is not required.
Let me know if there is better way and I'll update this part.

lslezak commented on README.md:
````diff
@@ -76,90 +85,74 @@ in clear text without any encryption (only converted to the Base64 encoding).
+
+For instance, to run `Table5.rb`, you can use following command:
+```
+YUI_HTTP_PORT=9999 Y2DIR=$PATH_TO_CLIENTS_DIR /usr/lib/YaST2/bin/y2base Table5 ncurses
````
There is a simpler way, run directly in the [yast-ycp-ui-bindings](https://github.com/yast/yast-ycp-ui-bindings) checkout (root not needed):

```shell
# Qt UI
YUI_HTTP_PORT=9999 /sbin/yast2 examples/Table5.rb
# ncurses UI
YUI_HTTP_PORT=9999 /sbin/yast examples/Table5.rb
```

rwx788 replied:
Oh, nice! I will update documentation. Thanks!

lslezak commented on README.md:
````diff
@@ -76,90 +85,74 @@ in clear text without any encryption (only converted to the Base64 encoding).
+In order to reduce the timeout system uses to unbind port, following command
+can be used to temporarily reduce it:
+```
+sudo sysctl -w sudo sysctl -w net.ipv4.tcp_fin_timeout=30
````
This changes the system behavior globally, I'm not sure which side effects or security issues this could have. I'd rather not touch it.

If needed we could add a new environment variable (e.g. `YUI_REUSE_PORT`) and use the value for setting the `MHD_OPTION_LISTENING_ADDRESS_REUSE` option.

rwx788 replied:
That sounds like a good idea, I will add new variable.

lslezak commented on API_v1.md:
```diff
@@ -1,19 +1,48 @@
-# LibYUI REST API v1
+<!-- TOC titleSize:2 tabSpaces:4 depthFrom:1 depthTo:6 withLinks:1 updateOnSave:1 orderedList:0 skip:0 title:1 charForUnorderedList:* -->
```
Um, how this was created and how it can be used?

rwx788 replied:
This was generated by plugin I have for table of content. I guess we can remove this.

lslezak replied:
I'm fine with that, I'm just interested in how can I regenerate TOC again when we add some new section...

Just add a comment what you need to install and how to regenerate it.

lslezak commented on src/YHttpWidgetsActionHandler.h:
```diff
@@ -43,45 +49,106 @@ class YHttpWidgetsActionHandler : public YHttpHandler
         }
         else {
             // TODO: demangle the C++ names here ?
             // https://gcc.gnu.org/onlinedocs/libstdc++/manual/ext_demangling.html
```
Please remove this comment as well...

lslezak commented on src/YHttpWidgetsActionHandler.h:
```diff
@@ -43,45 +49,106 @@ class YHttpWidgetsActionHandler : public YHttpHandler
+            YItem * item = selector->findItem( value );
+            if ( item )
+            {
+                // yuiMilestone() << "Activating item selector with item \"" << value << '"' << std::endl;
```
Either use `yuiDebug()` or remove the line. No dead code please...

lslezak commented on src/YHttpWidgetsActionHandler.h:
```diff
@@ -43,45 +49,106 @@ class YHttpWidgetsActionHandler : public YHttpHandler
+                selector->setKeyboardFocus();
+                // Toggle in case state selector undefined
+                bool select = state < 0  ? !item->selected() :
+                              state == 0 ? false :
```
It's pretty hard to read nested triple operators... Fortunately this part can be simplified to `state != 0`.

(The result of the comparison (`state == 0`) already is  a boolean, you do not need to convert it. Just change the condition to directly map to the result value...)

lslezak commented on src/YHttpWidgetsActionHandler.h:
```diff
@@ -43,45 +49,106 @@ class YHttpWidgetsActionHandler : public YHttpHandler
+                }
+                else
+                {
+                    select = state == 0 ? false : true;
```
The same here:

```suggestion
                    select = state =! 0;
```

lslezak commented on src/YHttpWidgetsActionHandler.cc:
```diff
@@ -239,71 +229,143 @@ int YHttpWidgetsActionHandler::do_action(YWidget *widget, const std::string &act
+            YComboBox* cb = dynamic_cast<YComboBox*>(widget);
+            if( !cb->editable() )
+            {
+                body << "{ \"error\" : \"Combobox '" << cb->label() << "' is not editable \" }" << std::endl;
```
Use the new `handle_error()` method also here.

lslezak commented on src/YHttpWidgetsActionHandler.cc:
```diff
@@ -239,71 +229,143 @@ int YHttpWidgetsActionHandler::do_action(YWidget *widget, const std::string &act
+            // Handle case when want select row by row number
+            return action_handler<YTable>( widget, body, [&] (YTable *tb) {
+
+                if( row_id >= tb->itemsCount() ) {
```
You can use [YTable::hasColumn()](https://github.com/libyui/libyui/blob/master/src/YTable.h#L95) directly.

rwx788 replied:
That would be useful for checking `column` parameter, but for the row we should check number of items. Or am I missing something?

lslezak commented on src/YHttpWidgetsActionHandler.cc:
```diff
@@ -239,71 +229,143 @@ int YHttpWidgetsActionHandler::do_action(YWidget *widget, const std::string &act
+                }
+                YItem * item = 0;
+                int current_row = 0;
+                for ( YItemConstIterator it = tb->itemsBegin(); it != tb->itemsEnd(); ++it )
```
The `YSelectionWidget` parent class implements the [`itemAt(int index)`](https://github.com/libyui/libyui/blob/master/src/YSelectionWidget.h#L359) method. Unfortunately it is protected and cannot be used form outside.

But as it would make sense for this code, maybe we should make it public... @shundhammer ?

shundhammer replied:
Yes, let's make it public. It would only be consistent. There is also itemsCount() which is also public, so having access to itemAt() would only be logical.

shundhammer replied:
Don't forget to adjust the BuildRequires libyui >= ... here when you change that!

rwx788 replied:
Alright, I will make relevant change in libyui. Thanks!

lslezak commented on src/YHttpWidgetsActionHandler.h:
```diff
@@ -43,45 +49,106 @@ class YHttpWidgetsActionHandler : public YHttpHandler
-                if (handler_func && widget->isEnabled()) handler_func(w);
+                if( !widget->isEnabled() )
+                {
+                    std::ostringstream error;
```
Um, do we really need `std::ostringstream` here? Simple `std::string` does not work here?

rwx788 replied:
I somehow decided to use stream here, as it performs better for concatenation. However we know the length of the string here, so you are right, we don't really need this here.
Thanks!

lslezak reviewed (commented):
Just some minor questions and suggestions...
Thanks a lot for the improvement!

Commit:
Fix item selection in YTable using YTableItem

For some reason, selection doesn't work for tables with icons if using
YItem. Casting to YTableItem resolves the issue.

See [poo#67393](https://progress.opensuse.org/issues/67393).

Commit:
Use itemAt for selection row in the Table

This requires libyui version 3.9.4.

We made itemAt method public for YSelectionWidget, so we can avoid
looping over items to select it by index.

Commit:
Add support for YCheckBoxFrame widget

Commit:
Add YUI_REUSE_PORT to allow reusing port

rwx788 commented:
@lslezak I've addressed all the review points. Please, take a look when you have time. Thanks for the support!

Commit:
Bump version to 0.5.0 and update changelog

Commit:
Add json serialization for YBarGraph

rwx788 reviewed (commented):
@lslezak Hey Ladi, so I've addressed all the points and added json serializtion for YBarGraph, so we can get all the changes. If you could review and we could get these changes merged, that would be great, so we could start preparing infrastructure for SP3. Thanks again for your help with this project!

jknphy commented on API_v1.md:
```diff
@@ -1,19 +1,46 @@
+        * [Parameters](#parameters)
+        * [Response](#response)
+        * [Examples](#examples)
+    * [Change Widgets, Do an Action</h3>](#change-widgets-do-an-actionh3)
```
we have this h3 tag does breaks the link.

rwx788 replied:
Fixed.

jknphy commented on API_v1.md:
```diff
@@ -123,39 +156,48 @@ Filter the widgets, one of:
 - **value** (optional) - new value or a parameter of the action
 - **column** (optional) - column id when selecting item in the table
-
+- **row** (optional) - row id when selecting item in the table
```
'row number' could be more descriptive? and similar for 'column name' above.

rwx788 replied:
Do you have some proposal? I don't see a need to add more details, unless we change the structure of the whole document which I was thinking to do.

jknphy replied:
Regardless the structure on the documents my comment was just related with not being ids or identificator, but maybe they can be called like that.

rwx788 replied:
Right, I though maybe stating that this value should be integer and call it row number instead of id, would make it more clear.

jknphy commented on API_v1.md:
```diff
@@ -123,39 +156,48 @@ Filter the widgets, one of:
   *value* parameter
-- **switch_radio** - activate radio button
 - **select** - select value in the combobox, row in the table or node in the
   tree, requires *value*parameter
```
s/*value*parameter/*value* parameter/

rwx788 replied:
Fixed.

jknphy commented on API_v1.md:
```diff
@@ -123,39 +156,48 @@ Filter the widgets, one of:
   - In case of table: select row in the table with given value. If
         *column* parameter is not provided, the first column will be used.  
   - In case of tree: select node in the tree. Use `|` as a delimiter for
         the child nodes.
```
I have to admit that this is difficult to phrase anyway. I"m not sure if it is clear without seen the example below.
Just a suggestion:

`Use '|' as a delimiter for the nodes composing the path to the node to be selected.`

Maybe it is not better, not sure :) depends on the reader.

rwx788 replied:
I will add an example, that is better than rephrasing, I guess =)

jknphy commented on API_v1.md:
```diff
@@ -123,39 +156,48 @@ Filter the widgets, one of:
 # select row with "test" cell value in the 2-nd column (counting from zero) in table with id "names"
 curl -X POST 'http://localhost:9999/v1/widgets?id=names&action=select&value=test&column=2'
 # select tree item with in tree with id "files"
 curl -X POST 'http://localhost:9999/v1/widgets?id=files&action=select&value=root|subnode|subnode'
```
Perhaps somewhere we could mention that we cannot select the control itself for now, only the nodes inside. Not sure if helps to write 'rootnode' instead of 'root' here but just want to mention I was not sure which root was referring, in case this could help someone else.

rwx788 replied:
What do you mean by selection control?

jknphy replied:
for example in yast2 partition we have System View (which I believe we cannot select) and inside there is a tree, which root the name of the machine, I got confused by that the tree was starting on System View. Ignore this if you don't see any confusing point.

rwx788 replied:
I mean, there is a possibility to set focus on the control, but yeah there are containers which have other objects inside of them. So I'm not sure if should explain that as well.

jknphy replied:
I think we don't have use case for clicking on them, so I guess it is fine how it is. We can see if someone encounters the same doubt root vs container in the future.

jknphy commented on API_v1.md:
````diff
@@ -123,39 +156,48 @@ Filter the widgets, one of:
 curl -X POST 'http://localhost:9999/v1/widgets?id=files&action=select&value=root|subnode|subnode'
-```
\ No newline at end of file
+# press url (<a href=\"firewall\">(enable)</a>) in richtext
````
Click embedded link in richtext?

jknphy commented on API_v1.md:
```diff
@@ -123,39 +156,48 @@ Filter the widgets, one of:
\ No newline at end of file
+# press url (<a href=\"firewall\">(enable)</a>) in richtext
+curl -X POST 'http://localhost:9999/v1/widgets?type=YRichText&action=select&value=firewall'
+# select menu item with label Image in parent item with label Document in YMenuButton menu
```
s/in parent item/(in|with) parent menu item/
s/YMenuButton/menubutton/ to follows same convention than above.

rwx788 replied:
SO we get `menubutton menu`? And which convention? I will put `menu button` if you think it helps with anything.

jknphy replied:
I just saw a pattern in lower case in the other example, in case we want to follow it, that is not used the name starting with 'Y'.

rwx788 replied:
Yeah, for some cases it's clear what tree is, in case of menu button, I would say it's a bit more libyui specific. So yeah, I guess overall it's better now.

jknphy commented on API_v1.md:
```diff
@@ -123,39 +156,48 @@ Filter the widgets, one of:
 - **enter_text** - set text in the field, the text is passed in the
   *value* parameter
-- **switch_radio** - activate radio button
 - **select** - select value in the combobox, row in the table or node in the
```
I guess we should mention richtext and menubutton and also below "in case of tree:" where menubutton also uses a list of nodes delimited by '|'.

rwx788 replied:
Good point. Done.

jknphy commented on README.md:
```diff
@@ -41,11 +57,15 @@ or (libyui-qt-rest-api)[https://github.com/libyui/libyui-ncurses-rest-api]).
+## Usage
+
+Many YaST modules require root privileges to run, the easiest way to start
+the application with rest API enabled, use the following commands:
```
s/ use the following/is (to use|using) the following/

rwx788 replied:
This is one of the sentences where I tried to re-phrase it so first part is from the new one and second from the old =D Fixed too.

jknphy commented on README.md:
````diff
@@ -76,90 +96,72 @@ in clear text without any encryption (only converted to the Base64 encoding).
+YUI_HTTP_PORT=9999 /sbin/yast2 examples/Table5.rb --ncurses
+```
+
+To run qt version of the app, simply replace `ncurses` parameter with `qt` as
````
s/run qt/run Qt/

rwx788 replied:
Fixed

jknphy commented on README.md:
```diff
@@ -41,11 +57,15 @@ or (libyui-qt-rest-api)[https://github.com/libyui/libyui-ncurses-rest-api]).
+Many YaST modules require root privileges to run, the easiest way to start
+the application with rest API enabled, use the following commands:
+* `su -c /bin/sh -c 'YUI_HTTP_PORT=9999 yast2 host --qt'` for Qt
+* `su -c /bin/sh -c 'YUI_HTTP_PORT=9999 yast2 host --ncurses` for ncurses.
```
s/--ncurses/--ncurses'/

rwx788 replied:
Nice catch

jknphy reviewed (commented):
Provided some minor comments about documentation.

shundhammer reviewed (approved):
LGTM
